File: legacy/src/gatling/scala/uk/co/baconi/oauth/api/feeders/Clients.scala
package uk.co.baconi.oauth.api.feeders

import io.gatling.core.Predef._
import io.gatling.core.session.{Expression, Session}

object Clients {

  private val CLIENT_ID = "clientId"
  private val CLIENT_SECRET = "clientSecret"

  object Expressions {

    val clientId: Expression[String] = session => session(CLIENT_ID).validate[String]

    val clientSecret: Expression[String] = session => session(CLIENT_SECRET).validate[String]

  }

  object Setup {

    def withClient(id: String, secret: String): Expression[Session] = session => session
      .set(CLIENT_ID, id)
      .set(CLIENT_SECRET, secret)

  }
}

File: automation/load/src/gatling/kotlin/uk/co/baconi/oauth/api/gatling/feeders/Clients.kt
package uk.co.baconi.oauth.api.gatling.feeders

import io.gatling.javaapi.core.Session
import uk.co.baconi.oauth.api.gatling.feeders.Clients.Client.Type.Confidential
import uk.co.baconi.oauth.api.gatling.feeders.Clients.Client.Type.Public

object Clients {

    private const val CLIENT = "client"

    data class Client(val id: String, val secret: String?, val redirect: String, val type: Type) {
        enum class Type {
            Confidential,
            Public;
        }
    }

    object Expressions {
        val client: (Session) -> Client = { session -> checkNotNull(session.get<Client>(CLIENT)) { "client was null" } }
        val clientId: (Session) -> String = { session -> client(session).id }
        val clientSecret: (Session) -> String? = { session -> client(session).secret }
        val clientRedirect: (Session) -> String = { session -> client(session).redirect }
    }

    object Setup {

        fun withConfidentialClient(id: String, secret: String, redirect: String): (Session) -> Session = { session ->
            session.set(CLIENT, Client(id, secret, redirect, Confidential))
        }

        fun withPublicClient(id: String, redirect: String): (Session) -> Session = { session ->
            session.set(CLIENT, Client(id, null, redirect, Public))
        }
    }
}